File: leetcode/cpp/PassedProblems/226-invert-binary-tree.cpp
/*
 * 226-invert-binary-tree.cpp
 *
 *  Created on: Sep 24, 2015
 *      Author: hanzhe
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#include <iostream>
using namespace std;
struct TreeNode {
      int val;
      TreeNode *left;
      TreeNode *right;
      TreeNode(int x) : val(x), left(NULL), right(NULL) {}
  };
class Solution {
public:
    TreeNode* invertTree(TreeNode* root) {
    	if(root == NULL)
    		return root;
    	TreeNode * l = root->left;
    	root->left = root->right;
    	root->right = l;
    	invertTree(root->left);
    	invertTree(root->right);
    	return root;
    }
};

File: leetcode/python/226-invert-binary-tree.py
#!/usr/bin/env python
# encoding: utf-8



"""

Invert a binary tree.

     4
   /   \
  2     7
 / \   / \
1   3 6   9
to
     4
   /   \
  7     2
 / \   / \
9   6 3   1
Trivia:
This problem was inspired by this original tweet by Max Howell:
Google: 90% of our engineers use the software you wrote (Homebrew), but you can’t invert a binary tree on a whiteboard so fuck off.
"""


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


# 2次过，简单
class Solution(object):
    def invertTree(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if root == None: return None
        l, r = root.left, root.right
        root.right = self.invertTree(l)
        root.left = self.invertTree(r)
        return root

